File: src/main/java/com/hiberus/anaya/redmineeditor/controller/Controller.java
package com.hiberus.anaya.redmineeditor.controller;

import com.hiberus.anaya.redmineeditor.model.ChangeEvents;
import com.hiberus.anaya.redmineeditor.model.Model;
import com.hiberus.anaya.redmineeditor.utils.FXUtils;
import javafx.application.Platform;
import javafx.util.Pair;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Consumer;

/**
 * The main application controller
 */
public class Controller {
    private final Model.Editor model = new Model.Editor(); // the model used

    private final List<Pair<Set<ChangeEvents>, ReadOnly>> listeners = new ArrayList<>(); // saved list of listeners and its data

    /**
     * A Runnable readonly model code
     * For foreground
     */
    public interface ReadOnly {
        /**
         * @param model readonly model
         */
        void run(Model model);
    }

    /**
     * A runnable for readwrite model code
     * For background
     */
    public interface ReadWrite {
        /**
         * @param model readwrite model
         * @throws MyException something went wrong
         */
        void run(Model.Editor model) throws MyException;
    }

    /**
     * Registers a new listener that will react to events.
     * Listeners will be called in register order
     *
     * @param events   list of events to react to
     * @param listener listener
     */
    public void register(Set<ChangeEvents> events, ReadOnly listener) {
        listeners.add(new Pair<>(events, listener));
    }

    /**
     * Run something in foreground.
     * Call this to have access to a readonly version of the model
     *
     * @param foreground what to run
     */
    public void runForeground(ReadOnly foreground) {
        foreground.run(model);
    }

    /**
     * Run something in background. Updates the loading status.
     * Call this to have access to a readWrite version of the model.
     *
     * @param background to run in background while the loading indicator is shown
     */
    public void runBackground(ReadWrite background) {
        runBackground(background, null);
    }

    /**
     * Run something in background, and then something in foreground. Updates the loading status.
     * Call this to have access to a readWrite version of the model.
     *
     * @param background to run in background while the loading indicator is shown
     * @param later      to run in foreground after the background thread finishes (the parameter indicates if there was an error)
     */
    public void runBackground(ReadWrite background, Consumer<Boolean> later) {
        // set as loading
        model.setLoading(true);
        fireChanges();

        // error container
        MyException[] error = new MyException[1];

        new Thread(() -> {
            try {
                // run in background
                background.run(model);
            } catch (MyException e) {
                error[0] = e; // save directly
            } catch (Throwable e) {
                // background error
                error[0] = new MyException("Internal error", "Something unexpected happened in background", e);
            }

            // unset as loading
            model.setLoading(false);
            fireChanges();

            Platform.runLater(() -> {
                // show error
                if (error[0] != null) error[0].showAndWait();
                // notify for later
                if (later != null) later.accept(error[0] == null);
            });
        }).start();
    }

    /**
     * Fires changes made to the model
     */
    public void fireChanges() {
        fireChanges(model.getChanges());
    }

    /**
     * Fires specific events
     *
     * @param events events to fire
     */
    public void fireChanges(Set<ChangeEvents> events) {
        listeners.forEach(pair -> {
            if (intersects(events, pair.getKey())) {
                // at least a registered event, run in foreground
                FXUtils.runInForeground(() -> pair.getValue().run(model));
            }
        });
    }

    /* ------------------------- utils ------------------------- */

    private static <E> boolean intersects(Set<E> a, Set<E> b) {
        // test if there is at least a common element in both sets
        HashSet<E> intersection = new HashSet<>(a);
        intersection.retainAll(b);
        return !intersection.isEmpty();
    }
}

File: src/main/java/com/hiberus/anaya/redmineeditor/controller/Controller.kt
package com.hiberus.anaya.redmineeditor.controller

import com.hiberus.anaya.redmineeditor.model.ChangeEvents
import com.hiberus.anaya.redmineeditor.model.Model
import com.hiberus.anaya.redmineeditor.settings.AppSettings
import com.hiberus.anaya.redmineeditor.utils.hiberus.LoadSpecialDays
import com.hiberus.anaya.redmineeditor.utils.runInForeground
import com.hiberus.anaya.redmineeditor.utils.stylize
import javafx.application.Platform
import javafx.scene.control.Alert
import javafx.stage.Window
import kotlin.concurrent.thread

// - Nooooo you need a whole 100TB of frameworks to use beans
// - haha, static go brrrrr
val AppController = Controller()

/**
 * The main application controller
 */
class Controller {

    /* ------------------------- data ------------------------- */

    /**
     * the model used
     */
    private val model = Model.Editor()

    /**
     * saved list of listeners and its data
     */
    private val listeners = mutableListOf<Pair<Set<ChangeEvents>, (Model) -> Unit>>()

    /* ------------------------- methods ------------------------- */

    /**
     * Registers a new listener that will react to events.
     * Listeners will be called in register order
     *
     * @param events   list of events to react to
     * @param listener listener
     */
    fun onChanges(events: Set<ChangeEvents>, listener: (Model) -> Unit) =
        listeners.add(events to listener)

    /**
     * Run something in foreground.
     * Call this to have access to a readonly version of the model
     *
     * @param foreground what to run
     */
    fun runForeground(foreground: (Model) -> Unit) = foreground(model)


    /**
     * Run something in background, and then something in foreground. Updates the loading status.
     * Call this to have access to a readWrite version of the model.
     *
     * @param background to run in background while the loading indicator is shown
     */
    fun runBackground(background: (Model.Editor) -> Unit) = runBackground(background) {}

    /**
     * Run something in background, and then something in foreground. Updates the loading status.
     * Call this to have access to a readWrite version of the model.
     *
     * @param background to run in background while the loading indicator is shown
     * @param later      to run in foreground after the background thread finishes (the parameter indicates if there was an error). Optional
     */
    fun runBackground(background: (Model.Editor) -> Unit, later: (Boolean) -> Unit = {}) {
        // set as loading
        model.isLoading = true
        fireChanges()

        thread(isDaemon = true) {
            val error = runCatching {
                // run in background
                background(model)
            }.exceptionOrNull()?.convert {
                // background error
                MyException("Internal error", "Something unexpected happened in background")
            }

            // unset as loading
            model.isLoading = false
            fireChanges()

            // then in foreground
            Platform.runLater {
                // show error
                error?.showAndWait()
                // notify for later
                later(error == null)
            }
        }
    }

    /**
     * Fires changes made to the model (or specific ones)
     *
     * @param events events to fire, those from the model by default
     */
    fun fireChanges(events: Set<ChangeEvents> = model.getChanges()) =
        listeners.also { println("Changes: $events") } // debug
            // get those who need to be modified
            .filter { (lEvents, _) -> lEvents intersects events }
            // and notify them all in foreground
            .let {
                runInForeground { // TODO: consider adding a try/catch here
                    it.forEach { (_, listener) -> listener(model) }
                }
            }

    /**
     * reloads the model data (loses changes, if existing)
     * Also reloads configuration if [reloadConfig] is set
     * Also resets the day if [resetDay] is set
     */
    fun reload(reloadConfig: Boolean = false, resetDay: Boolean = false) {
        var settingsERROR = false
        var specialDaysERROR = false
        runBackground({ model ->

            // reload files
            if (reloadConfig) {
                settingsERROR = !AppSettings.load()
                specialDaysERROR = !LoadSpecialDays()

                runInForeground {
                    // stylize displayed windows (should only be the main one)
                    Window.getWindows().map { it.scene }.distinct().forEach { it.stylize() }
                }
            }

            // set now
            if (resetDay) model.toNow()

            // reload data
            // TODO: don't reload when uploading, update internal state
            model.reloadRedmine(clearOnly = settingsERROR)

            // notify so that the ui is updated at this step and everything is updated
            AppController.fireChanges()

            // load month
            model.loadDate()

        }) {
            // after loading
            if (settingsERROR) {
                // invalid configuration, error
                Alert(Alert.AlertType.ERROR).apply {
                    title = "Configuration error"
                    contentText = "No valid configuration found"
                    stylize()
                }.showAndWait()
            }
            if (specialDaysERROR) {
                // invalid special days, warning
                Alert(Alert.AlertType.WARNING).apply {
                    title = "Special days error"
                    contentText = "No valid special days data found"
                    stylize()
                }.showAndWait()
            }
        }
    }

}

/* ------------------------- utils ------------------------- */

/**
 * test if there is at least a common element in both sets
 */
private infix fun <E> Set<E>.intersects(other: Set<E>) =
    (this intersect other).isNotEmpty()
